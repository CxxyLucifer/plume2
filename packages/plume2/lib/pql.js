"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var PartialQueryLang = /** @class */ (function () {
    function PartialQueryLang(lang) {
        this._lang = lang;
    }
    /**
     * 绑定bigQuery
     * 允许嵌套QL不允许嵌套PQL
     * @param bigQuery
     */
    PartialQueryLang.prototype.partialQL = function (bigQuery) {
        var _this = this;
        return function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            return bigQuery(_this._lang.apply(_this, args));
        };
    };
    return PartialQueryLang;
}());
exports.PartialQueryLang = PartialQueryLang;
exports.PQL = function (lang) {
    return new PartialQueryLang(lang);
};
