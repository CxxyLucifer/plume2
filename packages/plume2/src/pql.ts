/**
 * const helloPQL = PQL(index => QL([
 *   ['users', index, 'name'],
 *   (name)=>name
 * ])
 *
 * @Relax
 * class HellQL extends React.Component {
 *   static relaxProps = {
 *      name: helloPQL
 *   }
 *
 *   render() {
 *    const {name} = this.relaxProps;
 *    const username = name(1)
 *   }
 * }
 */
import { QueryLang } from './ql';

export class PartialQueryLang {
  constructor(lang: (...args: any[]) => QueryLang) {
    this._lang = lang;
  }

  private _lang: (...args: any[]) => QueryLang;

  /**
   * 绑定bigQuery
   * 允许嵌套QL不允许嵌套PQL
   * @param bigQuery
   */
  partialQL(bigQuery: Function) {
    return (...args: any[]) => {
      return bigQuery(this._lang(...args));
    };
  }
}

export const PQL = (lang: (...args: any[]) => QueryLang) => {
  return new PartialQueryLang(lang);
};
