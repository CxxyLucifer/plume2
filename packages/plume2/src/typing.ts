import { Map } from 'immutable';
import { ViewAction } from './view-action';

/**
 * Immutable的map类型
 */
export type IMap = Map<string, any>;

/**
 * actor的路由类型
 */
export type TRoute = {
  [name: string]: (state: IMap, params?: any) => IMap;
};

/**
 * Relax容器组件的上下文类型
 */
export interface IRelaxContext<Store> {
  _plume$Store: Store;
}

/**
 * Relax容器组件的props类型
 */
export interface IRelaxProps {
  relaxProps?: Object;
  [name: string]: any;
}

/**
 * Relax容器组件的类型
 */
export interface IRelaxComponent extends React.ComponentClass<IRelaxProps> {
  relaxProps?: Object;
}

/**
 * Store数据容器的参数配置类型
 */
export interface IOptions {
  debug?: boolean;
  [name: string]: any;
}

export interface IViewActionMapper {
  [name: string]: ViewAction;
}

export type TViewAction<T> = {
  [K in keyof T]: T[K] extends new (...args: Array<any>) => infer R ? R : any
};
