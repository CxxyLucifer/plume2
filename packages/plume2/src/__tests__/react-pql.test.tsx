import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { Map } from 'immutable';
import { Actor, Store, StoreProvider, QL, Relax } from '../index';
import { IMap } from '../typing';
import { PQL } from '../pql';

class ProductActor extends Actor {
  defaultState() {
    return {
      products: [
        { id: 1, name: 'p1' },
        { id: 2, name: 'p2' },
        { id: 3, name: 'p3' },
        { id: 4, name: 'p4' },
        { id: 5, name: 'p5' }
      ]
    };
  }
}

class AppStore extends Store {
  bindActor() {
    return [new ProductActor()];
  }
}

@StoreProvider(AppStore)
class ProductApp extends React.Component {
  store: Store;

  render() {
    const products = this.store.state().get('products');

    return (
      <div>
        {products.map((p, index) => {
          return <ProductItem key={p.get('id')} />;
        })}
      </div>
    );
  }
}

const productPQL = PQL(index => QL('productQL', [['products', index], p => p]));

@Relax
class ProductItem extends React.Component {
  static relaxProps = {
    productByIndex: productPQL
  };

  props: {
    relaxProps?: {
      productByIndex: (index: number) => IMap;
    };
  };

  render() {
    const { productByIndex } = this.props.relaxProps;
    const { id, name } = productByIndex(1).toJS();
    return (
      <div>
        <div>{id}</div>
        <div>{name}</div>
      </div>
    );
  }
}

describe('relax dql', () => {
  it('initialize', () => {
    const tree = renderer.create(<ProductApp />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
