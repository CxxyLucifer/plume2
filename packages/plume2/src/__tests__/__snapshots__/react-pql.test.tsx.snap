// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`relax dql initialize 1`] = `
<div>
  <div>
    <div>
      2
    </div>
    <div>
      p2
    </div>
  </div>
  <div>
    <div>
      2
    </div>
    <div>
      p2
    </div>
  </div>
  <div>
    <div>
      2
    </div>
    <div>
      p2
    </div>
  </div>
  <div>
    <div>
      2
    </div>
    <div>
      p2
    </div>
  </div>
  <div>
    <div>
      2
    </div>
    <div>
      p2
    </div>
  </div>
</div>
`;
